### PR TITLE
Np 1300 add search funtionality

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    public SearchResourcesResponse searchSingleTerm(String term) throws ApiGatewayException {
+
+        HttpRequest request = null;
+        try {
+            request = createHttpRequest(term);
+            HttpResponse<String> response = doSend(request);
+            logger.debug(response.body());
+            return toSearchResourcesResponse(response.body());
+        } catch (URISyntaxException | IOException | InterruptedException e) {
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:61]: I don't think the URISyntaxException is thrown

File: search-resources-api/src/main/java/no/unit/nva/search/RequestUtil.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Get searchTerm from request query parameters.
+     * @param requestInfo requestInfo
+     * @return searchTerm given in query parameter
+     * @throws ApiGatewayException exception containing explanatory message when parameter missing or inaccessible
+     */
+    public static String getSearchTerm(RequestInfo requestInfo) throws ApiGatewayException {
+
+        String searchTerm = null;
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/RequestUtil.java:27]: Initialise in try block

File: search-resources-api/src/test/java/no/unit/nva/search/ConstantsTest.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package no.unit.nva.search;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+public class ConstantsTest {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/ConstantsTest.java:7]: I don't think we need this the constants are likely tested elsewhere if they are in use

File: search-resources-api/src/test/java/RequestUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RequestUtilTest {
+
+    public static final String SAMPLE_QUERY_PARAMETER = "SampleQueryParameter";
+
+    @Test
+    public void canSearchTermRequest() throws ApiGatewayException {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/RequestUtilTest.java:15]: Is this not tested by handler testing?

File: search-resources-api/src/test/java/no/unit/nva/search/ElasticSearchRestClientTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ElasticSearchRestClientTest {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/ElasticSearchRestClientTest.java:20]: Functionality is likely tested in handler

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 6]
```diff
@@ -1,40 +1,59 @@
...
+        initEnvironment();
         searchResourcesApiHandler = new SearchResourcesApiHandler(environment);
     }
 
     @Test
-    public void testDefaultConstructor() {
-        SearchResourcesApiHandler searchResourcesApiHandler = new SearchResourcesApiHandler();
-        assertNotNull(searchResourcesApiHandler);
+    public void testDefaultConstructorFailsWhenNoEnvironmentDefined() {
+        assertThrows(IllegalStateException.class, () -> new SearchResourcesApiHandler());
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:47]: Replace with method reference

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 7]
```diff
@@ -1,40 +1,59 @@
...
-        environment = mock(Environment.class);
+        initEnvironment();
         searchResourcesApiHandler = new SearchResourcesApiHandler(environment);
     }
 
     @Test
-    public void testDefaultConstructor() {
-        SearchResourcesApiHandler searchResourcesApiHandler = new SearchResourcesApiHandler();
-        assertNotNull(searchResourcesApiHandler);
+    public void testDefaultConstructorFailsWhenNoEnvironmentDefined() {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:46]: DefaultConstructorThrowsIllegalStateExceptionWhenEnvironmentNotDefined

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 8]
```diff
@@ -1,40 +1,59 @@
...
+        assertThrows(IllegalStateException.class, () -> new SearchResourcesApiHandler());
     }
 
     @Test
     public void processInputReturnsNullWhenInputIsEmpty() throws ApiGatewayException {
         SearchResourcesRequest input = mock(SearchResourcesRequest.class);
         RequestInfo requestInfo = mock(RequestInfo.class);
         Context context = mock(Context.class);
-        assertThrows(InputException.class, () ->  searchResourcesApiHandler.processInput(input, requestInfo, context));
+        assertThrows(ApiGatewayException.class,
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:55]: This test does not throw an exception

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 9]
```diff
@@ -45,4 +64,24 @@ public void getSuccessStatusCodeReturnsOK() {
         assertEquals(statusCode, HttpStatus.SC_OK);
     }
 
+
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:67]: The naming of getSuccessStatusCodeReturnsOK makes me think that this test is tested elsewhere

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 10]
```diff
@@ -45,4 +64,24 @@ public void getSuccessStatusCodeReturnsOK() {
...
+    @Test
+    public void processInputReturnsSomething() throws ApiGatewayException, IOException, InterruptedException {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:69]: What does it return and when?

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java [hunk 11]
```diff
@@ -45,4 +64,24 @@ public void getSuccessStatusCodeReturnsOK() {
...
+        HttpResponse<String> httpResponse = mock(HttpResponse.class);
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:70]: Fix the unchecked assignment

Comment (brinxmat) [reply search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:70]: See code from other projects for examples

File: search-resources-api/src/test/java/no/unit/nva/search/exception/ExceptionsTest.java [hunk 12]
```diff
@@ -6,17 +6,26 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
-public class InputExceptionTest {
+public class ExceptionsTest {
```
Comment (brinxmat) [inline search-resources-api/src/test/java/no/unit/nva/search/exception/ExceptionsTest.java:9]: Test the exceptions from the handler using the Test utils library

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: search-resources-api/src/test/java/no/unit/nva/search/exception/ExceptionsTest.java (hunk 12 above)
Comment (svenngjeto) [reply search-resources-api/src/test/java/no/unit/nva/search/exception/ExceptionsTest.java:9]: Exception tested like the others

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java (hunk 7 above)
Comment (svenngjeto) [reply search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:46]: Renamed

File: search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java (hunk 6 above)
Comment (svenngjeto) [reply search-resources-api/src/test/java/no/unit/nva/search/SearchResourcesApiHandlerTest.java:47]: Code shortened

File: search-resources-api/src/test/java/RequestUtilTest.java (hunk 4 above)
Comment (svenngjeto) [reply search-resources-api/src/test/java/RequestUtilTest.java:15]: removed test as it was created for code coverage a long time ago (and test didn't d anythin useful anyway..)

File: search-resources-api/src/main/java/no/unit/nva/search/RequestUtil.java (hunk 2 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/RequestUtil.java:27]: Variable moved

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java (hunk 1 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:61]: removed the now unused exception

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    private SearchResourcesResponse toSearchResourcesResponse(String body) {
+        SearchResourcesResponse searchResourcesResponse = new SearchResourcesResponse();
+        logger.debug("Body={}", body);
+        return searchResourcesResponse;
+    }
+
+
+    private HttpRequest createHttpRequest(String term) throws URISyntaxException {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:73]: removed the now unthrown exception

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+        HttpRequest request = buildHttpRequest(term);
+
+        logger.debug(SEARCHING_LOG_MESSAGE, elasticSearchEndpointIndex, term);
+        return request;
+    }
+
+    private HttpRequest buildHttpRequest(String term) throws URISyntaxException {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:81]: done

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    /**
+     * Searches for an term or index:term in elasticsearch index.
+     * @param term search argument
+     * @throws ApiGatewayException thrown when uri is misconfigured, service i not available or interrupted
+     */
+    public SearchResourcesResponse searchSingleTerm(String term) throws ApiGatewayException {
+
+        HttpRequest request = null;
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:55]: Moved variable

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+import java.net.http.HttpResponse;
+
+public class ElasticSearchRestClient {
+
+
+    private static final Logger logger = LoggerFactory.getLogger(ElasticSearchRestClient.class);
+
+    public static final String INITIAL_LOG_MESSAGE = "using Elasticsearch endpoint {} {} and index {}";
+    public static final String SEARCHING_LOG_MESSAGE = "searching search index {}  for term {}";
+    public static final String SEARCH_SINGLE_TERM_LOG_MESSAGE = "searching for term {}";
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchRestClient.java:25]: Gone

File: search-resources-api/src/main/java/no/unit/nva/search/Constants.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package no.unit.nva.search;
+
+public class Constants {
+    public static final String CREATED_DATE_KEY = "createdDate";
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/Constants.java:4]: Gone

**Review by brinxmat — APPROVED** [review]

File: search-resources-api/src/main/java/no/unit/nva/search/SearchResourcesResponse.java [hunk 18]
```diff
@@ -1,32 +1,24 @@
...
-        }
-
-        public SearchResourcesResponse build() {
-            return new SearchResourcesResponse(this);
-        }
+@SuppressWarnings("PMD.ShortMethodName")
+public class SearchResourcesResponse extends ArrayList<JsonNode> {
+
+    /**
+     * Crates a list og strings containing hits from elasticsearch.
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/SearchResourcesResponse.java:14]: Creates

**Review by brinxmat — APPROVED** [review]

**Pull request merged**